### PR TITLE
Query : Adds string comparison alternative when converting LINQ to SQL

### LINKED ISSUE #2598: string.Compare not supported with LINQ to SQL
**Describe the bug**

[ExpressionToSQL::VisitBinary](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs#L454) does not support `string.Compare`, only `CompareTo`

We're using the [Microsoft.OData.Core](https://www.nuget.org/packages/Microsoft.OData.Core/7.9.0) nuget package to project OData queries into CosmosDB.

The Microsoft.OData.Core nuget package generates expressions with `string.Compare`

**To Reproduce**

We can reproduce without the OData dependency.

We have an `IQueryable` that fails with a `Where` clause like this:

```c#
.Where(m => string.Compare(m.Id, "hello") == 0)
```

However, it works if we do:

```c#
.Where(m => m.Id.CompareTo("hello") == 0);
```

**Expected behavior**

The SDK should generate valid SQL similar to 
```SQL
Id = "hello"
````

**Actual behavior**

The SDK throws an exception of type `DocumentQueryException` with the message "Method 'Compare' is not supported."

**Environment summary**
SDK Version: 3.20.1
OS Version (e.g. Windows, Linux, MacOSX): MacOS, Linux

**Additional context**

### PR DESCRIPTION
**Pull request by ernesto1596**
# Pull Request Template

## Description

Fixes issue #2598

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Closing issues

To automatically close an issue: closes #2598

### CODE REVIEW

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/LinqTranslationBaselineTests.cs [hunk 1]
```diff
@@ -1044,6 +1044,56 @@ public void TestStringCompareTo()
...
+                new LinqTestInput("Compare static string <=", b => getQuery(b).Select(doc => String.Compare(doc.StringField, "str") <= 0)),
+                // reverse operands
+                new LinqTestInput("Projected Compare == reverse operands", b => getQuery(b).Select(doc => 0 == String.Compare(doc.StringField, doc.StringField2))),
+                new LinqTestInput("Projected Compare < reverse operands", b => getQuery(b).Select(doc => 0 < String.Compare(doc.StringField, doc.StringField2))),
+                new LinqTestInput("Projected Compare <= reverse operands", b => getQuery(b).Select(doc => 0 <= String.Compare(doc.StringField, doc.StringField2))),
+                new LinqTestInput("Projected Compare > reverse operands", b => getQuery(b).Select(doc => 0 > String.Compare(doc.StringField, doc.StringField2))),
+                new LinqTestInput("Projected Compare >= reverse operands", b => getQuery(b).Select(doc => 0 >= String.Compare(doc.StringField, doc.StringField2))),
+                // errors Invalid compare value
+                new LinqTestInput("Compare > 1", b => getQuery(b).Select(doc => String.Compare(doc.StringField, "str") > 1)),
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/LinqTranslationBaselineTests.cs:1086]: For completeness, can you please add other error cases:
- unexpected number of arguments to string.Compare
- unexpected types of arguments to string.Compare
- unexpected types of arguments between string.Compare(a,b) and the constant
- non-constant argument

**Comment by adityasa** [comment]
>             }

A helper function for this + reuse would be nice.

---
Refers to: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:642 in 2d3d232. [](commit_id = 2d3d2325fe34c820f87caa55a627c99e2003490f, deletion_comment = False)

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs [hunk 2]
```diff
@@ -645,6 +649,76 @@ private static bool TryMatchStringCompareTo(MethodCallExpression left, ConstantE
...
+                        break;
+                    default:
+                        throw new DocumentQueryException(string.Format(CultureInfo.CurrentCulture, ClientResources.StringCompareToInvalidOperator));
+                }
+
+                // the constant value should be zero, otherwise we can't determine how to translate the expression
+                // it could be either integer or nullable integer
+                if (!(right.Type == typeof(int) && (int)right.Value == 0) &&
+                    !(right.Type == typeof(int?) && ((int?)right.Value).HasValue && ((int?)right.Value).Value == 0))
+                {
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:673]: How can this case be hit? Can you add coverage for this?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/BaselineTest/TestBaseline/LinqTranslationBaselineTests.TestStringCompareStatic.xml [hunk 3]
```diff
@@ -0,0 +1,257 @@
+﻿<Results>
+  <Result>
+    <Input>
+      <Description><![CDATA[Projected Compare ==]]></Description>
+      <Expression><![CDATA[query.Select(doc => (Compare(doc.StringField, doc.StringField2) == 0))]]></Expression>
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/BaselineTest/TestBaseline/LinqTranslationBaselineTests.TestStringCompareStatic.xml:5]: >Compare [](http://example.com/codeflow?start=48&length=7)

I am not sure why this says just "Compare" as opposed to "string.Compare", can you kindly double-check?

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs (hunk 2 above)
Comment (ernesto1596) [reply Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:673]: These are the test cases:
image

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/LinqTranslationBaselineTests.cs (hunk 1 above)
Comment (ernesto1596) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/LinqTranslationBaselineTests.cs]: - Added test for: unexpected number of arguments to string.Compare
- non-constant argument is already tested
- How do you propose to test this `unexpected types of arguments to string.Compare` and `unexpected types of arguments between string.Compare(a,b) and the constant`? Would that not be a compiler error when attempting to pass a non-string object into `String.Compare`?

File: Microsoft.Azure.Cosmos/src/ClientResources.Designer.cs [hunk 4]
```diff
@@ -582,6 +582,24 @@ internal class ClientResources {
             }
         }
         
+        /// <summary>
+        ///   Looks up a localized string similar to Invalid operator for string.Compare(). Valid operators are (&apos;==&apos;, &apos;&lt;&apos;, &apos;&lt;=&apos;, &apos;&gt;&apos; or &apos;&gt;=&apos;).
+        /// </summary>
+        internal static string StringCompareInvalidOperator {
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/ClientResources.Designer.cs:588]: Is this change intentional/necessary?

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs (hunk 2 above)
Comment (adityasa) [reply Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:680]: Thanks. I was curious about the Nullable int change more than non-nullable (for which I do see the coverage).

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/LinqTranslationBaselineTests.cs (hunk 1 above)
Comment (adityasa) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/LinqTranslationBaselineTests.cs]: Fair point, since there are no overloads of string.Compare that have different types of arguments. Thanks for following up.

File: Microsoft.Azure.Cosmos/src/ClientResources.Designer.cs (hunk 4 above)
Comment (ernesto1596) [reply Microsoft.Azure.Cosmos/src/ClientResources.Designer.cs:588]: It is intentional and I assume its necessary as otherwise it would use the StringCompareTo error strings.

**Review by adityasa — APPROVED** [review]
:shipit:

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs [hunk 5]
```diff
@@ -457,6 +457,10 @@ private static SqlScalarExpression VisitBinary(BinaryExpression inputExpression,
                 {
                     return ExpressionToSql.VisitStringCompareTo(methodCallExpression, constantExpression, inputExpression.NodeType, reverseNodeType, context);
                 }
+                if (TryMatchStringCompare(methodCallExpression, constantExpression, inputExpression.NodeType))
+                {
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:461]: newline

**Review by sboshra — DISMISSED** [review]
:shipit:

File: Microsoft.Azure.Cosmos/src/ClientResources.resx [hunk 6]
```diff
@@ -270,9 +270,15 @@
   <data name="StringCompareToInvalidConstant" xml:space="preserve">
     <value>The right hand side of string.CompareTo() comparison must be constant '0'</value>
   </data>
+  <data name="StringCompareInvalidConstant" xml:space="preserve">
+    <value>The right hand side of string.Compare() comparison must be constant '0'</value>
```
[the remark this answers is not in the document]
Comment (ernesto1596) [reply Microsoft.Azure.Cosmos/src/ClientResources.resx:274]: Updated

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs (hunk 5 above)
Comment (ernesto1596) [reply Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:461]: Updated

**Review by adityasa — DISMISSED** [review]
:shipit:

**Review by kirankumarkolli — DISMISSED** [review]
Approving on-behalf of Samer

**Review by adityasa — APPROVED** [review]
:shipit:

**Review by ealsur — APPROVED** [review]
